proposed/runtime-binding.md

```diff
@@ -0,0 +1,265 @@
+
+Let's look at an example.
+
+Requested version:
+
+* `2.1.0`
+
+Installed versions:
+
+* `1.1.17`
+* `3.0.0`
+* `3.0.1`
+* `3.1.0`
+* `4.0.0`
+
+In this scenario, the `3.0.1` runtime will be selected.
+
+## Handling Previews
+
+As of 3.0 Preview 2, previews have special binding behavior. In this proposal, previews are proposed to not have special behavior. Apps that depend on preview runtimes can bind to non-preview runtimes and vice-versa. Previews are not considered to be special going forward, with this proposal. This behavior is oriented on enabling applications built with previews to get a chance to run (likely successfully) on a stable version and to make testing stable applications on previews easy.
+
+Assumption: Stable Visual Studio versions only install stable .NET Core versions.
+
+## COM Components
+
+COM-based and other hosted environments have the requirement of loading managed components built for multiple runtime versions, and this set is typically not known a priori . 
+
+The application situation, described earlier in the doc, is largely oriented on missing runtime versions and how to handle that. The COM situation is oriented more on "best fit" for a set of components it may load. As a result, COM and other hosts need to artificially roll-forward to the latest acceptable runtime version to enable compatibility for arbitrary components.
+
+We will offer a configuration knobs for hosts that enable artificial roll-forward. These knobs play a similar role to [LockClrVersion](https://docs.microsoft.com/dotnet/framework/unmanaged-api/hosting/lockclrversion-function).
+
+## Configuration Knobs
+
+The host exposes the following configuration knobs that you can use to control binding behavior:
+
+### DOTNET_ROLL_FORWARD_ON_NO_CANDIDATE_FX
+
+This setting is described in [Roll Forward On No Candidate Fx](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/roll-forward-on-no-candidate-fx.md). It does not have the right behavior or UX. We will deprecate this setting for 3.0+.
+
+### RollForward
+
+We will introduce a new knob, called `RollForward`, with the following values:
+
+* `patch` -- Roll forward to the highest patch verison. This disables minor version roll forward.
+* `minor` -- Roll forward to the lowest higher minor verison, if requested minor version is missing.
+* `major` -- Roll forward to lowest higher major version, and lowest minor version, if requested major version is missing
+* `UseLatestMinor` -- Roll forward to latest minor version, even if requested minor version is on machine. Intended for COM hosting scenario.
+* `UseLatestMajor` -- Roll forward to latest major and highest minor version, even if requested major is on machine. Intended for COM hosting scenario.
```

Comment:
Nit 1: I think the values should all have the same canonical casing, so Patch, Minor, Major

Nit 2: I don't this Use is adding much value, how about LatestMinor, LatestMajor

Reply:
Good feedback. Let me share some of my thinking and see what you say. I original went with exactly what you have (ignoring the casing topic) `Minor` and `LatestMinor`. The problem is that the two settings are not symmetrical:

* `Minor` -- Do something only as a fallback case if a preferred runtime version is not found.
* `LatestMinor` -- Do something always.

That's why I added `Use` to `LatestMinor` to make it more obvious that this setting is more of a command than a request/hint.

I also thought of splitting these two forms of configuration into two different settings. That seemed like the worst idea, both because it would be confusing and there is no scenario where both settings would be set.

Thoughts?

Reply:
Maybe:
`AlwaysLatestMinor`
`AlwaysLatestMajor`